### PR TITLE
fix(gents): converts classes with inner typedefs to interfaces (1/4)

### PR DESCRIPTION
**Pull request by bowenni**
@rkirov @dpurp PTAL
Fixes https://github.com/angular/clutz/issues/451
Gents doesn't know how to handle classes with inner typedefs.
PR 1/4: TypeScript doesn't have the closure equivalent of nested type alias inside classes. So we convert them into top level interfaces and rename its use in the file.

### CODE REVIEW

File: src/test/java/com/google/javascript/gents/singleTests/inner_types.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+  /**
+   * equal function
+   * @export
+   */
+  equals(otherData: MyClass): boolean {
+    return this.type.a === otherData.type.a;
+  }
+}
+
+export type Type1 = {
```
Comment (dpurp) [inline src/test/java/com/google/javascript/gents/singleTests/inner_types.ts]: should these be converted into interfaces?

File: src/test/java/com/google/javascript/gents/singleTests/inner_types.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+MyClass.Type1 = {};
+
+/**
+ * @typedef {{
+ *     a: number,
+ * }}
+ */
+MyClass.Type2;
+
+exports = MyClass;
```
Comment (dpurp) [inline src/test/java/com/google/javascript/gents/singleTests/inner_types.js]: can you add a trailing newline?

File: src/main/java/com/google/javascript/gents/TypeConversionPass.java [hunk 3]
```diff
@@ -229,6 +261,42 @@ public void visit(NodeTraversal t, Node n, Node parent) {
     }
   }
 
+  /** Renames the inner typedefs to the new top level typedefs */
+  private class typeAliasConverter extends AbstractPostOrderCallback {
```
Comment (dpurp) [inline src/main/java/com/google/javascript/gents/TypeConversionPass.java]: - class name should be UpperCamelCase
- should this be static?

File: src/test/java/com/google/javascript/gents/multiTests/goog_provide_rewrite/goog_scope.js [hunk 4]
```diff
@@ -80,8 +80,10 @@ lorem.ipsum.baz = function() { return false; };
 
 
 /** @typedef {{key: number, value: string}} */
-Foo.InnerTypedef;
+Foo.InnerTypedef1;
 
+/** @typedef {{key: number, value: string}} */
+Foo.InnerTypedef2 = {};
```
Comment (gmoothart) [inline src/test/java/com/google/javascript/gents/multiTests/goog_provide_rewrite/goog_scope.js]: Can you give this a semantic name? Maybe `InnerTypedefWithAssignment`

File: src/test/java/com/google/javascript/gents/singleTests/inner_types.ts (hunk 1 above)
Comment (gmoothart) [inline src/test/java/com/google/javascript/gents/singleTests/inner_types.ts]: I don't know much about how closure handles these. But if they are public types, how will you handle references to them in other code (and can you write a multitest for that?). If they're private, you don't need to export.

Comment (bowenni) [inline src/test/java/com/google/javascript/gents/singleTests/inner_types.ts]: Acknowledged, will try to use interfaces instead.

File: src/test/java/com/google/javascript/gents/singleTests/inner_types.js (hunk 2 above)
Comment (bowenni) [inline src/test/java/com/google/javascript/gents/singleTests/inner_types.js]: Done.

File: src/main/java/com/google/javascript/gents/TypeConversionPass.java (hunk 3 above)
Comment (bowenni) [inline src/main/java/com/google/javascript/gents/TypeConversionPass.java]: Fixed the class name.

`TypeAliasConverter` is non-static because `typesToRename` from the enclosing class is used here.

File: src/main/java/com/google/javascript/gents/TypeConversionPass.java [hunk 5]
```diff
@@ -75,21 +82,46 @@ public void visit(NodeTraversal t, Node n, Node parent) {
...
               name = n.getSecondChild().getString();
             } else {
               name = n.getFirstChild().getString();
             }
             Node typeDef = Node.newString(Token.TYPE_ALIAS, name);
             types.put(name, typeDef);
             typeDef.setJSDocInfo(jsdoc);
+            if (needToExport) {
+              typesToRename.put(
+                  n.getFirstChild().getString() + "." + n.getSecondChild().getString(), name);
```
Comment (dpurp) [inline src/main/java/com/google/javascript/gents/TypeConversionPass.java]: is this different from `n.getQualifiedName()`?

File: src/test/java/com/google/javascript/gents/singleTests/inner_types.ts (hunk 1 above)
Comment (bowenni) [inline src/test/java/com/google/javascript/gents/singleTests/inner_types.ts]: Looked through some code in google3 and found that the references in other files are quite common.
It's currently not handled in this PR. With this PR only references in the same file are renamed.

I'm planning to work on a separate PR for that, along with some multitests.

File: src/test/java/com/google/javascript/gents/multiTests/goog_provide_rewrite/goog_scope.js (hunk 4 above)
Comment (bowenni) [inline src/test/java/com/google/javascript/gents/multiTests/goog_provide_rewrite/goog_scope.js]: Done.

File: src/main/java/com/google/javascript/gents/TypeConversionPass.java [hunk 6]
```diff
@@ -43,10 +46,14 @@ public void process(Node externs, Node root) {
       // We convert each file independently to avoid merging class methods from different files.
       if (child.isScript()) {
         this.types = new LinkedHashMap<>();
+        this.typesToRename = new HashMap<>();
```
Comment (rkirov) [inline src/main/java/com/google/javascript/gents/TypeConversionPass.java]: Use LinkedHashMap for iteration order that matches insertion order.

File: src/main/java/com/google/javascript/gents/TypeConversionPass.java (hunk 3 above)
Comment (rkirov) [inline src/main/java/com/google/javascript/gents/TypeConversionPass.java]: I wonder if it would better to extend TypeAnnotationConverter instead of a separate pass?

File: src/test/java/com/google/javascript/gents/singleTests/inner_types.ts (hunk 1 above)
Comment (bowenni) [inline src/test/java/com/google/javascript/gents/singleTests/inner_types.ts]: Converted to interfaces.
PTAL.

File: src/main/java/com/google/javascript/gents/TypeConversionPass.java (hunk 5 above)
Comment (bowenni) [inline src/main/java/com/google/javascript/gents/TypeConversionPass.java]: No they are the same.
I was overwhelmed by the number of available APIs and didn't find this one.

Fixed.

File: src/main/java/com/google/javascript/gents/TypeConversionPass.java (hunk 6 above)
Comment (bowenni) [inline src/main/java/com/google/javascript/gents/TypeConversionPass.java]: Done.

**Comment by bowenni** [comment]
PR 2 will be renaming references across multiple files and PR 3 will be converting function types in typedefs.

File: src/main/java/com/google/javascript/gents/TypeConversionPass.java [hunk 7]
```diff
@@ -229,6 +270,42 @@ public void visit(NodeTraversal t, Node n, Node parent) {
     }
   }
 
+  /** Renames the inner typedefs to the new top level typedefs */
```
Comment (rkirov) [inline src/main/java/com/google/javascript/gents/TypeConversionPass.java]: Nothing typedef specific in this class. Describe what the class does first and then what it is used for in the grand scheme on things. The way I read this class is isolation - it renames everything in jsDOCs based on typesToRename map.

File: src/test/java/com/google/javascript/gents/singleTests/inner_types.js [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+/**
+ * @typedef {{
+ *     a: {b: {c: number}},
+ * }}
+ */
+MyClass.InnerTypedefWithNestedTypes;
+
+/**
+ * @typedef {{
```
Comment (rkirov) [inline src/test/java/com/google/javascript/gents/singleTests/inner_types.js]: We can consider down the road to change those to interfaces too, now that we have the capabilities.

File: src/test/java/com/google/javascript/gents/singleTests/inner_types.js [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+ *     a: number,
+ * }}
+ */
+MyClass.InnerTypedef;
+
+/**
+ * @typedef {{
+ *     a: {b: {c: number}},
```
Comment (rkirov) [inline src/test/java/com/google/javascript/gents/singleTests/inner_types.js]: add second (non-nested) member too

File: src/test/java/com/google/javascript/gents/singleTests/inner_types.js (hunk 8 above)
Comment (bowenni) [inline src/test/java/com/google/javascript/gents/singleTests/inner_types.js]: Convert all `@typedef`s to interfaces and `@type` to type aliases?

File: src/main/java/com/google/javascript/gents/GentsCodeGenerator.java [hunk 10]
```diff
@@ -146,6 +146,18 @@ boolean maybeOverrideCodeGen(Node n) {
           return true;
         }
         return false;
+      case EXPORT:
+        // When a type alias is exported, closure code generator will add two semi-colons, one for type alias and one for export
+        // For example: export type T = {key: string};;
+        if (1 != n.getChildCount()) {
```
Comment (bowenni) [inline src/main/java/com/google/javascript/gents/GentsCodeGenerator.java]: Done.

File: src/main/java/com/google/javascript/gents/TypeConversionPass.java [hunk 11]
```diff
@@ -69,36 +77,69 @@ public void visit(NodeTraversal t, Node n, Node parent) {
           }
           break;
         case GETPROP:
+          // Converts an inner typedef into a top level interface, which then later has its members
+          // converted in TypeAnnotationPass. Don't convert it into a type alias because TypeScript
+          // doesn't have nested type aliases inside a class.
```
Comment (bowenni) [inline src/main/java/com/google/javascript/gents/TypeConversionPass.java]: Done.

File: src/main/java/com/google/javascript/gents/TypeConversionPass.java (hunk 7 above)
Comment (bowenni) [inline src/main/java/com/google/javascript/gents/TypeConversionPass.java]: Done.

File: src/test/java/com/google/javascript/gents/singleTests/inner_types.js (hunk 9 above)
Comment (bowenni) [inline src/test/java/com/google/javascript/gents/singleTests/inner_types.js]: Done.

File: src/test/java/com/google/javascript/gents/singleTests/inner_types.js (hunk 8 above)
Comment (rkirov) [inline src/test/java/com/google/javascript/gents/singleTests/inner_types.js]: just the first part

Comment (bowenni) [inline src/test/java/com/google/javascript/gents/singleTests/inner_types.js]: Acknowledged, will do a separate PR.

**Comment by dpurp** [comment]
Awesome! Thank you!
